doc/dev/rule-streaming.md

```diff
@@ -0,0 +1,241 @@
+production of rules to a callback that it obtains via
+`Build_config`. Inside Dune, this callback is implemented by the
+`Gen_rules` module inside the `dune_rules` library. `dune_rules` is
+the library that is responsible for parsing, interpreting and
+compiling `dune` files down to low-level build rules.
+
+### `Dune_rules.Gen_rules`
+
+The entry of `Dune_rules.Gen_rules` is the `gen_rules` function. It's
+API looks like:
+
+```ocaml
+val gen_rules : dir:Path.Build.t -> Dune_engine.Rules.t Memo.Build.t
+```
+
+Where `Dune_rules.Rules.t` is a "raw" set of rules. Raw in the sense
+that there is no overlap checks or any other checks. During the rule
+production phase, we merely accumulate a set of rules that is later
+processed. The API of `gen_rules` is in fact a bit more complex, but
+the above definition is enough for the purpose of this document.
+
+The first thing `gen_rules` do is analyse the directory it is
+given. If the directory correspond to a source directory with a `dune`
+file, `gen_rules` will dispatch the call to the part of `dune_rules`
+that parses and interprets the `dune` file. This is the simplest case,
+but even in this case there are some things worth mentioning.
+
+For instance, when compiling an OCaml library dune stores the
+artifacts for the library in generated dot-directories. For instance,
+the cmi files for library `foo` living in source directory `src` will
+end up in `_build/default/src/.foo.objs/byte`. We could produce these
+rules when `gen_rules` is called with direcotry
+`_build/default/src/.foo.objs/byte`, however that would spread out the
+logic for interpreting `library` stanzas. It is much simpler to
+produce all the build rules corresponding to a `library` stanza in one
+go. This is what is hapenning at the moment: when called with
+directory `_build/default/src`, `gen_rules` will not only produce
+rules for this directory but will also produce rules for
+`_build/default/src/.foo.objs/byte` and various other directories.
+
+`Load_rules` doesn't know anything about this. And in particular, it
+doesn't know that it is the `gen_rules` call for directory
+`_build/default/src/` that will produce the rules for the dot
+subdirectories. When `Load_dir` loads the rules for the
+`.../.foo.objs/byte` sub-directory, it simply calls `gen_rules` with
+this directory. It is `gen_rules` that "redirects" the call to the
+`_build/default/src` directory by calling
+`Load_rules.load_dir_and_produce_its_rules`. This function simply
+calls `Load_rules.load_dir` and re-emits all the raw rules that were
+returned by the corresponding `gen_rules` call.
+
+This works because `Load_rules.load_dir` accepts the facts that
+`gen_rules` produces rules for many directory at once. It simply
+filters out the result. But for things to behave well, the unwritten
+following invariant must hold: `gen_rules ~dir:d` is allowed to
+generate rules for directory `d'` iff `gen_rules ~dir:d'` emits a call
+to `Load_rules.load_dir ~dir:d`.
```

Comment:
At this point of reading I thought that the current behaviour and invariants are pretty crazy in terms of complexity :)

Reply:
Agreed! I've been tightening the the rule production API since last night and found a lot of odd stuff:

- in many places we just keep generating the same rules again and again
- in some places we produce rules that will never be looked at

I've added checks and I'm almost done fixing the various places where we do these things.

Reply:
Yep, just looked at the PRs -- nice clean up!

Reply:
And just put the last one up: https://github.com/ocaml/dune/pull/5270

Reply:
👍